### PR TITLE
go-blocking: Increase threads for connections

### PR DESCRIPTION
**Pull request by methane**

### CODE REVIEW

**Comment by jonhoo** [comment]
I'll run a benchmark with this patch later today, as well as one without `LockOSThread` as suggested by @diegobernardes. I'll also try one with `GOMAXPROCS(201)` to see what kind of overhead increasing `GOMAXPROCS` on the fly has.

**Comment by jonhoo** [comment]
go locking comparison

It's hard to say which approach is better. Incrementing `GOMAXPROCS` on the fly seems like it provides some performance boost, but given that the hardcoded value of 201 is too high when the number of clients is less than 200, this might be misleading. `LockOSThread` appears to help performance somewhat, but its performance also drops when the number of cores grows large. I'm not entirely sure why this happens.

**Comment by diegobernardes** [comment]
Nice to see two things in the chart:
- With `incthreads` and `fixthreads` the latency has been reduced drastically
- The speedup curve is better with lock

So golang now has more or less the same performance as rust without the multiplex, lets see how golang-multiplex performs.

The real test gonna happen when someone implement rust with mio.

---

I asked in [gonuts](https://groups.google.com/forum/#!topic/golang-nuts/3GnRkCIflDA) group about the change of GOMAXPROCS during runtime and Dave Cheney answered this:

> It will be both. Goroutines that return to the scheduler will find that their hosting thread is surplus to requirements and they will be reassigned to another threads work pool. However, work in progress will not be affected until those goroutines block on a syscall or channel send/ receive.

I really don't know if this can impact or not the performance :disappointed:

**Comment by jonhoo** [comment]
I'm pretty sure dynamically _increasing_ `GOMAXPROCS` shouldn't make much of a difference. _Decreasing_ might carry a performance penalty, but we're not doing that here.

Good catch about the latency. You're right, this seems to have helped significantly, though that is probably just because more threads can now block simultaneously. I'm still concerned that the right-hand side of the performance curve flattens with this change though, whereas the fixed-thread implementation does not.

I'm waiting on a response from @methane before I can test #23, but yes, it will be interesting to see. I'm not quite as convinced it is a reasonable design though, as it is essentially a less flexible implementation of multiplexing (just like `rust-multiplex`).

File: servers/go-blocking/main.go [hunk 1]
```diff
@@ -38,17 +46,29 @@ func main() {
 			log.Println("failed to accept connection:", err)
 			continue
 		}
+		nc := atomic.AddInt32(&numConns, 1)
+		if int(nc) >= numThreads {
+			numThreads *= 2
```
Comment (diegobernardes) [inline servers/go-blocking/main.go:51]: Why `*= 2` ?
In the case of 200 concurrent connections we gonna set the GOMAXPROCS to 400, the double of c/rust.
Or am i wrong?

**Comment by jonhoo** [comment]
@diegobernardes: my guess is that this was added to amortize the cost (if any) of incrementing `GOMAXPROCS`. Since it is double each time, we will only have to do `log n` increments instead of `n` for `n` clients.

**Comment by diegobernardes** [comment]
> I'm pretty sure dynamically increasing GOMAXPROCS shouldn't make much of a difference. Decreasing might carry a performance penalty, but we're not doing that here.

You're right, don't know why i was thinking the code was decreasing the GOMAXPROCS.

> Good catch about the latency. You're right, this seems to have helped significantly, though that is probably just because more threads can now block simultaneously. I'm still concerned that the right-hand side of the performance curve flattens with this change though, whereas the fixed-thread implementation does not.

Don't have any clue of what that can be.
The only thing that i suspect is in [this line](https://github.com/jonhoo/volley/pull/24/files#r32958413).
The chart show less spikes and more smooth line in the lock version, maybe there is something related to this line.

File: servers/go-blocking/main.go (hunk 1 above)
Comment (methane) [inline servers/go-blocking/main.go:51]: GOMAXPROCS is not number of threads, it's just _maximum_ number of threads to run goroutine.
(o/t There are another few threads).

So when number of client is 200, only 200 or less threads are running for the connection.

**Pull request merged**